File: bin/add-localization.py
#!/usr/bin/env python3
"""
bin/add-localization.py — Add German localization entries to Localizable.xcstrings.

Usage:
    # Single entry
    bin/add-localization.py "Settings" "Einstellungen"
    bin/add-localization.py "Target interval: %@" "Zielintervall: %@"

    # Batch from JSON file (object with "key": "German translation")
    bin/add-localization.py --batch translations.json

    # Batch from CSV (key,translation per line, no header)
    bin/add-localization.py --batch translations.csv

    # List existing keys (useful for checking what's already there)
    bin/add-localization.py --list
    bin/add-localization.py --list | grep -i interval

    # Show entries missing a German translation
    bin/add-localization.py --missing

    # Dry run (show what would be added/changed without writing)
    bin/add-localization.py --dry-run "Settings" "Einstellungen"
    bin/add-localization.py --dry-run --batch translations.json

Options:
    --xcstrings PATH    Path to .xcstrings file (default: auto-detect in project)
    --batch FILE        Read translations from JSON or CSV file
    --list              List all existing localization keys
    --missing           Show keys without a German translation
    --dry-run           Show changes without writing
    --sort              Sort keys alphabetically when writing (default: preserve order)
"""

import argparse
import csv
import json
import os
import sys
from pathlib import Path


def find_xcstrings():
    """Auto-detect the Localizable.xcstrings file."""
    # Try common locations relative to the script or cwd
    candidates = [
        Path("Peach/Resources/Localizable.xcstrings"),
        Path("Resources/Localizable.xcstrings"),
        Path("Localizable.xcstrings"),
    ]

    # Also try relative to the script location
    script_dir = Path(__file__).resolve().parent.parent
    for c in list(candidates):
        candidates.append(script_dir / c)

    for c in candidates:
        if c.exists():
            return c

    # Search upward from cwd
    cwd = Path.cwd()
    for parent in [cwd] + list(cwd.parents):
        found = list(parent.rglob("Localizable.xcstrings"))
        if found:
            return found[0]

    return None


def load_xcstrings(path):
    with open(path, 'r', encoding='utf-8') as f:
        return json.load(f)


def save_xcstrings(path, data, sort_keys=False):
    strings = data.get('strings', {})
    if sort_keys:
        data['strings'] = dict(sorted(strings.items()))

    with open(path, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)
        f.write('\n')


def make_entry(german_value):
    """Create a localization entry for a German translation."""
    return {
        "localizations": {
            "de": {
                "stringUnit": {
                    "state": "translated",
                    "value": german_value
                }
            }
        }
    }


def add_translations(data, translations, dry_run=False):
    """Add translations, return (added, updated, skipped) counts."""
    strings = data.setdefault('strings', {})
    added = 0
    updated = 0
    skipped = 0

    for key, german in translations.items():
        if key in strings:
            # Check if German translation already exists and matches
            existing_de = (strings[key]
                          .get('localizations', {})
                          .get('de', {})
                          .get('stringUnit', {})
                          .get('value'))

            if existing_de == german:
                skipped += 1
                if dry_run:
                    print(f"  skip (identical): {key!r}")
                continue

            # Update existing entry, preserving other localizations
            if 'localizations' not in strings[key]:
                strings[key]['localizations'] = {}
            strings[key]['localizations']['de'] = {
                'stringUnit': {
                    'state': 'translated',
                    'value': german
                }
            }
            updated += 1
            if dry_run:
                old = existing_de or "(none)"
                print(f"  update: {key!r}: {old!r} → {german!r}")
        else:
            strings[key] = make_entry(german)
            added += 1
            if dry_run:
                print(f"  add: {key!r} → {german!r}")

    return added, updated, skipped


def load_batch_file(path):
    """Load translations from a JSON or CSV file."""
    p = Path(path)

    if p.suffix == '.json':
        with open(p, 'r', encoding='utf-8') as f:
            data = json.load(f)
        if not isinstance(data, dict):
            print(f"Error: JSON file must be an object with {{\"key\": \"translation\"}}",
                  file=sys.stderr)
            sys.exit(1)
        return data

    # Assume CSV (also handles .tsv, .txt)
    translations = {}
    with open(p, 'r', encoding='utf-8') as f:
        # Auto-detect delimiter
        sample = f.read(4096)
        f.seek(0)
        if '\t' in sample:
            delimiter = '\t'
        else:
            delimiter = ','

        reader = csv.reader(f, delimiter=delimiter)
        for row in reader:
            if len(row) >= 2:
                key = row[0].strip()
                value = row[1].strip()
                if key and value:
                    translations[key] = value

    return translations


def list_keys(data):
    strings = data.get('strings', {})
    for key in sorted(strings.keys()):
        de = (strings[key]
              .get('localizations', {})
              .get('de', {})
              .get('stringUnit', {})
              .get('value', ''))
        if de:
            print(f"{key}  →  {de}")
        else:
            print(f"{key}  →  (no German translation)")


def show_missing(data):
    strings = data.get('strings', {})
    count = 0
    for key in sorted(strings.keys()):
        de = (strings[key]
              .get('localizations', {})
              .get('de', {})
              .get('stringUnit', {})
              .get('value'))
        if not de:
            print(key)
            count += 1
    print(f"\n{count} keys missing German translation", file=sys.stderr)


def main():
    parser = argparse.ArgumentParser(
        description='Add German localizations to Localizable.xcstrings')
    parser.add_argument('key', nargs='?', help='Localization key (English string)')
    parser.add_argument('german', nargs='?', help='German translation')
    parser.add_argument('--xcstrings', help='Path to .xcstrings file')
    parser.add_argument('--batch', metavar='FILE',
                        help='Read translations from JSON or CSV file')
    parser.add_argument('--list', action='store_true',
                        help='List all existing keys and their German translations')
    parser.add_argument('--missing', action='store_true',
                        help='Show keys without German translations')
    parser.add_argument('--dry-run', action='store_true',
                        help='Show what would change without writing')
    parser.add_argument('--sort', action='store_true',
                        help='Sort keys alphabetically when writing')

    args = parser.parse_args()

    # Find the xcstrings file
    if args.xcstrings:
        xcstrings_path = Path(args.xcstrings)
    else:
        xcstrings_path = find_xcstrings()

    if not xcstrings_path or not xcstrings_path.exists():
        print("Error: Could not find Localizable.xcstrings", file=sys.stderr)
        print("Specify path with --xcstrings", file=sys.stderr)
        sys.exit(1)

    data = load_xcstrings(xcstrings_path)

    # --- List mode ---
    if args.list:
        list_keys(data)
        return

    # --- Missing mode ---
    if args.missing:
        show_missing(data)
        return

    # --- Collect translations ---
    translations = {}

    if args.batch:
        translations = load_batch_file(args.batch)
    elif args.key and args.german:
        translations = {args.key: args.german}
    else:
        parser.print_help()
        sys.exit(1)

    if not translations:
        print("No translations to add.", file=sys.stderr)
        sys.exit(1)

    # --- Apply ---
    if args.dry_run:
        print(f"Dry run against {xcstrings_path}:\n")

    added, updated, skipped = add_translations(data, translations, dry_run=args.dry_run)

    if not args.dry_run:
        save_xcstrings(xcstrings_path, data, sort_keys=args.sort)

    action = "Would add" if args.dry_run else "Added"
    print(f"\n{action} {added}, updated {updated}, skipped {skipped} "
          f"(of {len(translations)} entries) in {xcstrings_path.name}")


if __name__ == '__main__':
    main()

File: bin/add-localization.swift
#!/usr/bin/env swift
//
// bin/add-localization.swift — Add German localization entries to Localizable.xcstrings.
//
// Usage:
//   # Single entry
//   bin/add-localization.swift "Settings" "Einstellungen"
//
//   # Batch from JSON file (object with "key": "German translation")
//   bin/add-localization.swift --batch translations.json
//
//   # Batch from CSV (key,translation per line, no header)
//   bin/add-localization.swift --batch translations.csv
//
//   # List existing keys
//   bin/add-localization.swift --list
//
//   # Show entries missing a German translation
//   bin/add-localization.swift --missing
//
//   # Dry run
//   bin/add-localization.swift --dry-run "Settings" "Einstellungen"
//
// Options:
//   --xcstrings PATH    Path to .xcstrings file (default: auto-detect)
//   --batch FILE        Read translations from JSON or CSV file
//   --list              List all existing keys and German translations
//   --missing           Show keys without German translations
//   --dry-run           Show changes without writing

import Foundation

// MARK: - File Discovery

func findXCStrings() -> URL? {
    let fileManager = FileManager.default
    let cwd = URL(fileURLWithPath: fileManager.currentDirectoryPath)

    let candidates = [
        "Peach/Resources/Localizable.xcstrings",
        "Resources/Localizable.xcstrings",
        "Localizable.xcstrings",
    ]

    // Try relative to cwd
    for candidate in candidates {
        let url = cwd.appendingPathComponent(candidate)
        if fileManager.fileExists(atPath: url.path) {
            return url
        }
    }

    // Try relative to script location
    let scriptDir = URL(fileURLWithPath: #filePath).deletingLastPathComponent().deletingLastPathComponent()
    for candidate in candidates {
        let url = scriptDir.appendingPathComponent(candidate)
        if fileManager.fileExists(atPath: url.path) {
            return url
        }
    }

    return nil
}

// MARK: - Xcode-Compatible JSON Encoding

func xcodeJSONEncode(_ value: Any, indent: Int = 0) -> String {
    let prefix = String(repeating: " ", count: indent)

    if let dict = value as? [String: Any] {
        if dict.isEmpty {
            return "{\n\n\(prefix)}"
        }
        let sortedKeys = dict.keys.sorted { $0.localizedStandardCompare($1) == .orderedAscending }
        var items: [String] = []
        for key in sortedKeys {
            let encodedKey = jsonEncodeString(key)
            let encodedValue = xcodeJSONEncode(dict[key]!, indent: indent + 2)
            items.append("\(prefix)  \(encodedKey) : \(encodedValue)")
        }
        return "{\n" + items.joined(separator: ",\n") + "\n\(prefix)}"
    }

    if let array = value as? [Any] {
        if array.isEmpty {
            return "[]"
        }
        var items: [String] = []
        for element in array {
            let encodedValue = xcodeJSONEncode(element, indent: indent + 2)
            items.append("\(prefix)  \(encodedValue)")
        }
        return "[\n" + items.joined(separator: ",\n") + "\n\(prefix)]"
    }

    if let string = value as? String {
        return jsonEncodeString(string)
    }

    if let number = value as? NSNumber {
        if number === kCFBooleanTrue {
            return "true"
        }
        if number === kCFBooleanFalse {
            return "false"
        }
        if number is Int {
            return "\(number.intValue)"
        }
        return "\(number.doubleValue)"
    }

    if value is NSNull {
        return "null"
    }

    return "\(value)"
}

func jsonEncodeString(_ string: String) -> String {
    var result = "\""
    for char in string.unicodeScalars {
        switch char {
        case "\"": result += "\\\""
        case "\\": result += "\\\\"
        case "\n": result += "\\n"
        case "\r": result += "\\r"
        case "\t": result += "\\t"
        case "\u{08}": result += "\\b"
        case "\u{0C}": result += "\\f"
        default:
            if char.value < 0x20 {
                result += String(format: "\\u%04x", char.value)
            } else {
                result += String(char)
            }
        }
    }
    result += "\""
    return result
}

// MARK: - XCStrings I/O

func loadXCStrings(from url: URL) throws -> [String: Any] {
    let data = try Data(contentsOf: url)
    guard let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        throw NSError(domain: "add-localization", code: 1, userInfo: [
            NSLocalizedDescriptionKey: "Invalid xcstrings format: root must be a JSON object"
        ])
    }
    return dict
}

func saveXCStrings(_ data: [String: Any], to url: URL) throws {
    let output = xcodeJSONEncode(data)
    try output.write(to: url, atomically: true, encoding: .utf8)
}

// MARK: - Translation Operations

func getGermanValue(from entry: [String: Any]) -> String? {
    guard let localizations = entry["localizations"] as? [String: Any],
          let de = localizations["de"] as? [String: Any],
          let stringUnit = de["stringUnit"] as? [String: Any],
          let value = stringUnit["value"] as? String else {
        return nil
    }
    return value
}

func makeEntry(germanValue: String) -> [String: Any] {
    [
        "localizations": [
            "de": [
                "stringUnit": [
                    "state": "translated",
                    "value": germanValue,
                ]
            ]
        ]
    ]
}

struct TranslationResult {
    var added = 0
    var updated = 0
    var skipped = 0
}

func addTranslations(
    to data: inout [String: Any],
    translations: [(key: String, german: String)],
    dryRun: Bool
) -> TranslationResult {
    var strings = data["strings"] as? [String: Any] ?? [:]
    var result = TranslationResult()

    for (key, german) in translations {
        if var entry = strings[key] as? [String: Any] {
            let existingDE = getGermanValue(from: entry)

            if existingDE == german {
                result.skipped += 1
                if dryRun {
                    print("  skip (identical): \(quoted(key))")
                }
                continue
            }

            // Update existing entry, preserving other localizations
            var localizations = entry["localizations"] as? [String: Any] ?? [:]
            localizations["de"] = [
                "stringUnit": [
                    "state": "translated",
                    "value": german,
                ]
            ]
            entry["localizations"] = localizations
            strings[key] = entry
            result.updated += 1
            if dryRun {
                let old = existingDE ?? "(none)"
                print("  update: \(quoted(key)): \(quoted(old)) → \(quoted(german))")
            }
        } else {
            strings[key] = makeEntry(germanValue: german)
            result.added += 1
            if dryRun {
                print("  add: \(quoted(key)) → \(quoted(german))")
            }
        }
    }

    data["strings"] = strings
    return result
}

// MARK: - Batch Loading

func parseCSVLine(_ line: String, delimiter: Character) -> [String] {
    var fields: [String] = []
    var current = ""
    var inQuotes = false
    var i = line.startIndex

    while i < line.endIndex {
        let char = line[i]
        if inQuotes {
            if char == "\"" {
                let next = line.index(after: i)
                if next < line.endIndex, line[next] == "\"" {
                    current.append("\"")
                    i = line.index(after: next)
                } else {
                    inQuotes = false
                    i = line.index(after: i)
                }
            } else {
                current.append(char)
                i = line.index(after: i)
            }
        } else if char == "\"" {
            inQuotes = true
            i = line.index(after: i)
        } else if char == delimiter {
            fields.append(current)
            current = ""
            i = line.index(after: i)
        } else {
            current.append(char)
            i = line.index(after: i)
        }
    }
    fields.append(current)
    return fields
}

func loadBatchFile(from url: URL) throws -> [(key: String, german: String)] {
    if url.pathExtension == "json" {
        let data = try Data(contentsOf: url)
        guard let dict = try JSONSerialization.jsonObject(with: data) as? [String: String] else {
            throw NSError(domain: "add-localization", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "JSON file must be an object with {\"key\": \"translation\"}"
            ])
        }
        return dict.map { ($0.key, $0.value) }
    }

    // CSV/TSV — supports RFC 4180 quoted fields
    let content = try String(contentsOf: url, encoding: .utf8)
    let delimiter: Character = content.contains("\t") ? "\t" : ","
    var translations: [(String, String)] = []

    for line in content.components(separatedBy: .newlines) {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { continue }

        let fields = parseCSVLine(trimmed, delimiter: delimiter)
        guard fields.count >= 2 else { continue }

        let key = fields[0].trimmingCharacters(in: .whitespaces)
        let value = fields[1].trimmingCharacters(in: .whitespaces)
        guard !key.isEmpty, !value.isEmpty else { continue }
        translations.append((key, value))
    }

    return translations
}

// MARK: - List / Missing

func listKeys(in data: [String: Any]) {
    guard let strings = data["strings"] as? [String: Any] else { return }
    let sortedKeys = strings.keys.sorted { $0.localizedStandardCompare($1) == .orderedAscending }

    for key in sortedKeys {
        if let entry = strings[key] as? [String: Any],
           let de = getGermanValue(from: entry) {
            print("\(key)  →  \(de)")
        } else {
            print("\(key)  →  (no German translation)")
        }
    }
}

func showMissing(in data: [String: Any]) {
    guard let strings = data["strings"] as? [String: Any] else { return }
    let sortedKeys = strings.keys.sorted { $0.localizedStandardCompare($1) == .orderedAscending }
    var count = 0

    for key in sortedKeys {
        if let entry = strings[key] as? [String: Any] {
            if getGermanValue(from: entry) == nil {
                print(key)
                count += 1
            }
        } else {
            print(key)
            count += 1
        }
    }

    printErr("\n\(count) keys missing German translation")
}

// MARK: - Helpers

func printErr(_ message: String) {
    FileHandle.standardError.write(Data((message + "\n").utf8))
}

func quoted(_ s: String) -> String {
    "'\(s)'"
}

func exitWithError(_ message: String) -> Never {
    printErr("Error: \(message)")
    exit(1)
}

// MARK: - Argument Parsing

struct Arguments {
    var xcstringsPath: String?
    var batchFile: String?
    var list = false
    var missing = false
    var dryRun = false
    var key: String?
    var german: String?
}

func parseArguments() -> Arguments {
    var args = Arguments()
    let argv = Array(CommandLine.arguments.dropFirst())
    var i = 0

    while i < argv.count {
        switch argv[i] {
        case "--xcstrings":
            i += 1
            guard i < argv.count else { exitWithError("--xcstrings requires a path") }
            args.xcstringsPath = argv[i]
        case "--batch":
            i += 1
            guard i < argv.count else { exitWithError("--batch requires a file path") }
            args.batchFile = argv[i]
        case "--list":
            args.list = true
        case "--missing":
            args.missing = true
        case "--dry-run":
            args.dryRun = true
        case "--help", "-h":
            printUsage()
            exit(0)
        default:
            if argv[i].hasPrefix("-") {
                exitWithError("Unknown option: \(argv[i])")
            }
            if args.key == nil {
                args.key = argv[i]
            } else if args.german == nil {
                args.german = argv[i]
            } else {
                exitWithError("Too many positional arguments")
            }
        }
        i += 1
    }

    return args
}

func printUsage() {
    print("""
    Usage: add-localization.swift [OPTIONS] [KEY GERMAN]

    Add German localization entries to Localizable.xcstrings.

    Arguments:
      KEY                 Localization key (English string)
      GERMAN              German translation

    Options:
      --xcstrings PATH    Path to .xcstrings file (default: auto-detect)
      --batch FILE        Read translations from JSON or CSV file
      --list              List all existing keys and German translations
      --missing           Show keys without German translations
      --dry-run           Show changes without writing
      -h, --help          Show this help
    """)
}

// MARK: - Main

let args = parseArguments()

// Find xcstrings file
let xcstringsURL: URL
if let path = args.xcstringsPath {
    xcstringsURL = URL(fileURLWithPath: path)
} else if let found = findXCStrings() {
    xcstringsURL = found
} else {
    exitWithError("Could not find Localizable.xcstrings. Specify path with --xcstrings")
}

guard FileManager.default.fileExists(atPath: xcstringsURL.path) else {
    exitWithError("File not found: \(xcstringsURL.path)")
}

var data: [String: Any]
do {
    data = try loadXCStrings(from: xcstringsURL)
} catch {
    exitWithError("Failed to load \(xcstringsURL.lastPathComponent): \(error.localizedDescription)")
}

// List mode
if args.list {
    listKeys(in: data)
    exit(0)
}

// Missing mode
if args.missing {
    showMissing(in: data)
    exit(0)
}

// Collect translations
var translations: [(key: String, german: String)] = []

if let batchFile = args.batchFile {
    if args.key != nil {
        exitWithError("Cannot use --batch together with positional arguments.")
    }
    do {
        translations = try loadBatchFile(from: URL(fileURLWithPath: batchFile))
    } catch {
        exitWithError("Failed to load batch file: \(error.localizedDescription)")
    }
} else if let key = args.key, let german = args.german {
    translations = [(key, german)]
} else if args.key != nil {
    exitWithError("Missing German translation argument.")
} else {
    printUsage()
    exit(1)
}

guard !translations.isEmpty else {
    exitWithError("No translations to add.")
}

// Apply
if args.dryRun {
    print("Dry run against \(xcstringsURL.lastPathComponent):\n")
}

let result = addTranslations(to: &data, translations: translations, dryRun: args.dryRun)

if !args.dryRun {
    do {
        try saveXCStrings(data, to: xcstringsURL)
    } catch {
        exitWithError("Failed to write \(xcstringsURL.lastPathComponent): \(error.localizedDescription)")
    }
}

let action = args.dryRun ? "Would add" : "Added"
let total = translations.count
print("\n\(action) \(result.added), updated \(result.updated), skipped \(result.skipped) (of \(total) entries) in \(xcstringsURL.lastPathComponent)")
